lib/mail-archive-helper.ts

```diff
@@ -172,10 +173,13 @@ export class MailArchiveGitHelper {
             if (line.startsWith("@@ ")) {
                 const match = line.match(/^@@ -(\d+,)?\d+ \+(\d+,)?(\d+)?/);
                 if (match) {
+                    if (counter) {
+                        console.log(`Oops: unprocessed buffer ${buffer}`);
+                    }
                     counter = parseInt(match[3], 10);
                     buffer = "";
                 }
-            } else if (counter) {
+            } else if (counter && line.match(/^[ +]/)) {
```

Comment:
Not sure that `counter` matters anymore.  You are in a patch or not in a patch.  A patch starts with the `@@` line and continues with lines `/^[-+ ]/`.  Using counter will have problems when there are common lines after the last insert and decrementing it for common lines will cut off the email.  See [this code around line 170](https://github.com/webstech/gitgitgadget/commit/8708b800ccbc963e4b91a59381b35dfe0e30d1d2) where the patch state is terminated by a character that is not the patch prefix.  It also handles the last buffer at the end.

Reply:
> Using counter will have problems when there are common lines after the last insert and decrementing it for common lines will cut off the email.

That should not be the case. The `+` range in the `@@` line includes the context (or "common") lines. Otherwise the diff is corrupt.

> Not sure that `counter` matters anymore. You are in a patch or not in a patch.

I wanted to make sure that I work on a file by file basis ;-) That's why I parse the `@@` header and then act as soon as the hunk is complete: the expectation is that the hunk encompasses the complete file.

And I would like to keep this code, as it would probably even handle future changes to the layout where multiple mails might be added at once (not that I expect such a change to come, but still, I like the code to be prepared to handle even that).

Reply:
> The + range in the @@ line includes the context (or "common") lines.

My mistake.  That makes my changes just a different implementation.

fyi, commit c9379c29a7493780a2f1d39b3f14a12d867d5ecf in the lore repo has two diffs.  The first one deletes all the content and the second one adds the new content.  There are a few of them like that.

Reply:
> commit c9379c29a7493780a2f1d39b3f14a12d867d5ecf in the lore repo has two diffs. The first one deletes all the content and the second one adds the new content. There are a few of them like that.

Interesting! I had not realized that. It removes the file `d` and adds the file `m`... 

I guess the reason is fb163402d41dbf62a3b336b3f6f39cb6fe56c9e8 which marks `m` as spam. In the previous layout, those spam mails would be removed from the resulting tree... I guess it is a downside of the new layout that you cannot remove spam mails effectively (because the mails are no longer supposed to live in one huge top-level tree but purely in the Git object database instead).